File: Stack/9012_bracket.py
import sys

for bracket in range(int(sys.stdin.readline())):
    bracket = list(sys.stdin.readline().strip())
    # count 설정
    count = 0
    for idx in range(len(bracket)):
        # 만약 ( 이면 +1, )이면 -1
        if bracket[idx] == '(':
            count += 1
        elif bracket[idx] == ')':
            count -= 1
        # 음수가 되면  NO, for 문도 정지
        if count < 0:
            print('NO')
            break    
    else:
        # 다 돌았을때 숫자가 같으면(중간에 음수가 되는 것 없이) YES
        if count == 0:
            print('YES')
        else:
            print('NO')

File: Stack/9012_bracket.js
// 9012 괄호
const fs = require("fs");

const settings = () => {
    // const input = fs.readFileSync("input.txt").toString().trim().split("\n");
    const input = fs.readFileSync("/dev/stdin").toString().split("\n");
    return input.slice(1);
}


const main = (bracket) => {
    let stack = 0;
    let ans = "YES"
    for (let b of bracket) {
        if (b === "("){
            stack ++;
        } else {
            stack --;
        }
        // console.log(b, stack)
        if (stack < 0){
            ans = "NO";
            break;
        }
    }
    // console.log(stack.length);

    if (stack !== 0) {
        ans = "NO";
    }

    console.log(ans);
}

const brackets = settings();

for (let bracket of brackets) {
    main(bracket);
}

